spirv_msl.cpp

```diff
@@ -15196,6 +15235,133 @@ void CompilerMSL::analyze_argument_buffers()
 
+// Return the resource type of the app-provided resources for the descriptor set,
+// that matches the resource index of the argument buffer index.
+MSLResourceBinding& CompilerMSL::get_argument_buffer_resource(uint32_t desc_set, uint32_t arg_idx)
+{
+	for (auto itr = resource_bindings.begin(); itr != resource_bindings.end(); itr++)
+	{
+		auto& rez_bind = itr->second.first;
+		uint32_t rez_idx = ~0;
+		switch (rez_bind.base_type)
+		{
+			case SPIRType::Void:
+				rez_idx = rez_bind.msl_buffer;
+				break;
+			case SPIRType::Image:
+			case SPIRType::SampledImage:
+				rez_idx = rez_bind.msl_texture;
+				break;
+			case SPIRType::Sampler:
+				rez_idx = rez_bind.msl_sampler;
+				break;
+			default:
+				SPIRV_CROSS_THROW("Unexpected argument buffer resource base type. When padding argument buffer elements, all descriptor set resources must be supplied with a base type by the app.");
+				break;
+		}
+
+		if (rez_bind.desc_set == desc_set && rez_idx == arg_idx)
+			return rez_bind;
+	}
+	SPIRV_CROSS_THROW("Argument buffer resource base type could not be determined. When padding argument buffer elements, all descriptor set resources must be supplied with a base type by the app.");
+	static MSLResourceBinding unkwn_rez;
+	return unkwn_rez;
+}
+
+// Adds an argument buffer padding argument buffer type as one or more members of the struct type at the member index.
+// Metal does not support arrays of buffers, so these are emitted as multiple struct members.
+void CompilerMSL::add_argument_buffer_padding_buffer_type(SPIRType& struct_type, uint32_t& mbr_idx, uint32_t count)
+{
+	if (!argument_buffer_padding_buffer_type_id)
+	{
+		uint32_t buff_type_id = ir.increase_bound_by(2);
+		auto &buff_type = set<SPIRType>(buff_type_id);
+		buff_type.basetype = SPIRType::Void;
+		buff_type.storage = StorageClassUniformConstant;
+
+		uint32_t ptr_type_id = buff_type_id + 1;
+		auto &ptr_type = set<SPIRType>(ptr_type_id);
+		ptr_type = buff_type;
+		ptr_type.pointer = true;
+		ptr_type.pointer_depth++;
+		ptr_type.parent_type = buff_type_id;
+
+		argument_buffer_padding_buffer_type_id = ptr_type_id;
+	}
+	for (uint32_t rez_idx = 0; rez_idx < count; rez_idx++) {
+		add_argument_buffer_padding_type(argument_buffer_padding_buffer_type_id, struct_type, mbr_idx, 1);
+	}
+}
+
+// Adds an argument buffer padding argument image type as a member of the struct type at the member index.
+void CompilerMSL::add_argument_buffer_padding_image_type(SPIRType& struct_type, uint32_t& mbr_idx, uint32_t count)
+{
+	if (!argument_buffer_padding_image_type_id)
+	{
+		uint32_t base_type_id = ir.increase_bound_by(2);
+		auto &base_type = set<SPIRType>(base_type_id);
+		base_type.basetype = SPIRType::Float;
+		base_type.width = 32;
+
+		uint32_t img_type_id = base_type_id + 1;
+		auto &img_type = set<SPIRType>(img_type_id);
+		img_type.basetype = SPIRType::Image;
+		img_type.storage = StorageClassUniformConstant;
+
+		img_type.image.type = base_type_id;
+		img_type.image.dim = Dim2D;
```

Comment:
Does this actually work? In Metal, the size of an image descriptor can change based on state like dimension, MSAA, etc. Afaik, the type must match exactly.

Reply:
>Does this actually work?

This definitely works on all the macOS devices I've tested (AMD, Intel, M1). On macOS, all argument descriptors are the same size (8 bytes). And all app testing with this design I've done has been successful so far on macOS.

However, this is flaky on iOS, and I'm still working through it. 

>In Metal, the size of an image descriptor can change based on state like dimension, MSAA, etc. Afaik, the type must match exactly.

What is the source of this statement? Is this documented somewhere that I haven't found yet, or is it an assumption based on the requirement for `MTLTextureType` as part of the argument encoder configuration?

On iOS, buffer, image, and sampler arg descriptors have different sizes from one another, but measurements I've done by asking `[MTLArgumentEncoder encodedLength]` produces a consistent 128 bytes for all `MTLTextureTypes`. 

Based on that, I've been working under the assumption that the need for `MTLTextureType` was not related to size specifically, but encoding. In that case, after taking into consideration resource type, a standard padding space for image args independent of `MTLTextureType` _**should**_ work. But I must admit I haven't got this working entirely properly yet on iOS, and perhaps size does matter here after all (even though it is not obviously reported through the size measurement API).

The advantages of getting this approach working wherever possible are compelling, because it associates the Metal argument buffer with a descriptor set, rather than a pipeline shader stage, and separates descriptor set updates from pipeline encoding, as Vulkan intends. Once written to, a descriptor set Metal argument buffer can be bound to multiple pipelines across multiple pipeline stages, without having to repeatedly dynamically write to a separate Metal argument buffer for each pipeline stage.

So...I'd like to include this approach as an option to enable on platforms like macOS that will support it. On iOS, we'll see, and on iOS we might have to turn this off and swap in some form of dynamic writing to per-pipeline-stage Metal argument buffers. Very non-Vulkan and very inefficient. Ugggh!

@cdavis5e You did a chunk of work on this earlier, and any insight you have would be appreciated as well.

Reply:
> even though it is not obviously reported through the size measurement API

128 seems excessive. I'm going to assume this is some kind of alignment. Probably try taking the size of an array of 1024 images of different types?

Reply:
To be clear, I'm not against this PR. If you can ship it in MoltenVK, then it's good enough for me.

Reply:
> `MTLTextureType` as part of the argument encoder configuration?

I recall in portability discussions that you had to add in texture type in Vulkan somehow to be able to use IAB, but my memory is hazy.

Reply:
>128 seems excessive. I'm going to assume this is some kind of alignment. Probably try taking the size of an array of 1024 images of different types?

It does seem excessive, doesn't it? But it seems to be real on iOS:

```
[mvk-info] Metal arg buff for 1024 Buffer storage is 8200.
[mvk-info] Metal arg buff for 1024 Sampler storage is 16392.
[mvk-info] Metal arg buff for 1024 MTLTextureType1D storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType1DArray storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType2D storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DArray storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DMultisample storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DMultisampleArray storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeCube storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeCubeArray storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureType3D storage is 131080.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeTextureBuffer storage is 131080.
```

Each arg buffer has an 8-byte "header" or equivalent. Other than that, the remaining content is 128 bytes per for textures.

Contrast with macOS. In this case, Apple Silicon M1, which make the above even more odd, since the architectures are similar:

```
[mvk-info] Metal arg buff for 1024 Buffer storage is 8192.
[mvk-info] Metal arg buff for 1024 Sampler storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType1D storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType1DArray storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType2D storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DArray storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DMultisample storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType2DMultisampleArray storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeCube storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeCubeArray storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureType3D storage is 8192.
[mvk-info] Metal arg buff for 1024 MTLTextureTypeTextureBuffer storage is 8192.
```
No "header" and 8-bytes per resource descriptor, regardless of type.

Reply:
> This definitely works on all the macOS devices I've tested (AMD, Intel, M1). On macOS, all argument descriptors are the same size (8 bytes). And all app testing with this design I've done has been successful so far on macOS.
> 
> > In Metal, the size of an image descriptor can change based on state like dimension, MSAA, etc. Afaik, the type must match exactly.
> 
> What is the source of this statement?

Me. When *I* tested it on my AMD Radeon Pro 460, buffer descriptors were 16 bytes; and texture descriptors were 32 bytes, except for 2D MS, which was 64 bytes--but strangely not 2D MS array, which was again 32 bytes.

Maybe it depends on your OS version? I tested on Mojave. That you tested on M1 suggests you are using Big Sur. Maybe they're actually working on my request to make this agnostic of texture type?

Reply:
> > This definitely works on all the macOS devices I've tested (AMD, Intel, M1). On macOS, all argument descriptors are the same size (8 bytes). And all app testing with this design I've done has been successful so far on macOS.
> > > In Metal, the size of an image descriptor can change based on state like dimension, MSAA, etc. Afaik, the type must match exactly.
> > 
> > 
> > What is the source of this statement?
> 
> Me. When _I_ tested it on my AMD Radeon Pro 460, buffer descriptors were 16 bytes; and texture descriptors were 32 bytes, except for 2D MS, which was 64 bytes--but strangely not 2D MS array, which was again 32 bytes.
> 
> Maybe it depends on your OS version? I tested on Mojave. That you tested on M1 suggests you are using Big Sur. Maybe they're actually working on my request to make this agnostic of texture type?

It occurs to me that you tested multiple descriptors per buffer, but I only tested one. You found that the buffer had an 8-byte header, because the descriptor count doesn't go evenly into the buffer size. That might account for some of the different sizes we saw.

Reply:
So I tried it with multiple elements this time, both with `arrayLength` greater than one, and with multiple descriptors... and I found the same results as before: buffers are 16 bytes, textures are 32 bytes, 2D MS textures (but not 2D MS array) are 64 bytes. I tried samplers, too (I didn't last time); they are also 16 bytes.

Yeah, I think it depends on your OS version.

Reply:
>Maybe it depends on your OS version? I tested on Mojave. That you tested on M1 suggests you are using Big Sur. Maybe they're actually working on my request to make this agnostic of texture type?

Yes. Big Sur in all cases. You might be right about the OS advances changing things. On macOS, all descriptors, independent of type, are 8 bytes now. And for textures and samplers those seem to be just numeric indexes into some kind of lookup. Perhaps they are caching the actual descriptors themselves somewhere else, and not in the actual `MTLBuffer` used as the argument buffer?

And odd that iOS has such a huge size for texture descriptors. Assuming macOS M1 hardware would be similar to iOS, that's another indication that some kind of reference abstraction might now be happening within the driver on macOS when it comes to arg buffer content.

Reply:
I also tried it on my Intel GPU on Mojave. There I see the same results as @billhollings: buffers, textures, and samplers are all 8 bytes. Are you sure you tested using the AMD GPU? I wonder if Apple has started doing this for all GPUs because programs have begun to depend on this behavior.

Reply:
I also tried it on my M1 on Big Sur, and I get the same results as Bill there, too.

Reply:
Yes...I tested all three macOS GPU's I have (AMD, Intel, and M1). AMD is actually the default on the machine I run CTS on.

Maybe they had to create an abstraction for Intel, and are now standardizing it to all drivers in order to make it consistent and not so dependent on types (per your filed request).

And...don't get me started on the requirement to specify usage for every resource on every command encoder. That almost completely negates the call reduction gained by submitting the resources in an argument buffer.